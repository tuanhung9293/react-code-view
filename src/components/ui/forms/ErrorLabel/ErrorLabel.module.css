@value colors: 'styles/colors.css';
@value error from colors;
@value font-sizes: 'styles/font-sizes.css';
@value small from font-sizes;

.base {
  font-size: small;
  color: error;
}