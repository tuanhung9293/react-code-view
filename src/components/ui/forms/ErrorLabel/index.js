import React from 'react'
import styles from './ErrorLabel.module.css'

function ErrorLabel({...rest}) {
  return <label className={styles.base} {...rest} />
}

export default ErrorLabel