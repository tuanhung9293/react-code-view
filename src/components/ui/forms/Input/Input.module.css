@value colors: 'styles/colors.css';
@value separator, field, error, secondary from colors;

.input{
  appearance: none;
  display: block;
  border: 2px solid transparent;
  background-color: field;
  border-radius: 4px;
  box-shadow: 0 2px rgba(0,0,0,0.1);
  margin: 10px 0;
  padding: 10px 15px;
  font-family: 'campton', Helvetica, sans-serif;
  width: 100%;
  min-width: 260px;
  box-sizing: border-box;
}

.error{
  border-color: error;
}

.input[type=text]:focus,
.input[type=email]:focus,
.input[type=password]:focus,
.input[type=date]:focus{
  outline: 0;
  border-color: secondary;
}