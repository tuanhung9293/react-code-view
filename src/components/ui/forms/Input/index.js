import React from 'react'
import classNames from 'classnames'
import styles from './Input.module.css'

function Input({type, hasError, value, ...rest}) {
  const inputValue = value || ''
  const className = classNames({
    [styles.input]: true,
    [styles.error]: hasError
  })
  return (
    <input className={className} type={type || 'text'} value={inputValue} {...rest} />
  )
}

export default Input