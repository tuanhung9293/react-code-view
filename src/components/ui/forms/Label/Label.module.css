.base{
  font-size: 0.875rem;
}

.title {
  font-size: 33px;
}

.base, .title{
  font-weight: bold;
  display: block;
}