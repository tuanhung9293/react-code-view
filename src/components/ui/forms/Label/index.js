import React from 'react'
import classNames from 'classnames'
import styles from './Label.module.css'

function Label({children, title}) {
  const className = classNames({
    [styles.base]: true,
    [styles.title]: title
  })
  return (
    <label className={className}>{children}</label>
  )
}

export default Label