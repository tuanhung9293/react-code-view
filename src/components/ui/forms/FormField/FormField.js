import React from 'react'
import PropTypes from 'prop-types'
import Label from '../Label'
import ErrorLabel from '../ErrorLabel'
import Input from '../Input'
import debounce from 'lodash/debounce'
import styles from './FormField.module.css'

class FormField extends React.PureComponent {
  debouncedValidate = null
  componentDidMount() {
    const {name, validate} = this.props
    if (validate) {
      this.context._form && this.context._form.registerValidator(name, validate)
      this.debouncedValidate = debounce(this.validate, 50)
      this.debouncedValidate()
    }
  }

  componentWillUnmount() {
    const {name} = this.props
    this.context._form && this.context._form.unregisterValidator(name)
    if (this.debouncedValidate) {
      this.debouncedValidate.cancel()
    }
  }

  render() {
    const {
      formName,
      label,
      touched,
      validationError,
      component: Component,
      onBlur,
      onValidate,
      name,
      afterChange,
      ...rest
    } = this.props
    const displayError = validationError && touched ? validationError : null
    const componentProps = {
      name,
      ...rest,
      onBlur: this.handleBlur,
      onChange: this.handleChange
    }
    delete componentProps.validate
    delete componentProps.validateField
    delete componentProps.onValidationError
    return (
      <div className={styles.wrapper}>
        <div className={styles.labels}>
          <Label>{label}</Label>
          <ErrorLabel>
            {displayError}
          </ErrorLabel>
        </div>
        {Component ? (
          <Component {...componentProps} />
        ) : (
          <Input type="text" {...componentProps} />
        )}
      </div>
    )
  }

  handleChange = e => {
    const {onChange, afterChange, validate} = this.props
    if (validate && this.debouncedValidate) {
      this.debouncedValidate()
    }
    if (onChange) {
      onChange(e)
    }
    if (afterChange) {
      afterChange(e)
    }
  }

  handleBlur = e => {
    if (this.props.validate) {
      this.validate()
    }
    this.props.onBlur(e)
  }

  handleValidationError = error => {
    if (error && this.props.onValidationError) {
      this.props.onValidationError(error)
    }
  }

  validate = () => {
    const {value, validate: validator, validateField} = this.props
    const error = validator(value)
    if (error) {
      this.handleValidationError(error)
    }
    validateField(error)
  }
}

FormField.propTypes = {
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  component: PropTypes.func,
  validate: PropTypes.func
}

FormField.contextTypes = {
  _form: PropTypes.object
}

export default FormField
