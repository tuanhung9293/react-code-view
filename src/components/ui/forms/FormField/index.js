import React from 'react'
import {connect} from 'react-redux'
import PropTypes from 'prop-types'
import FormField from './FormField'
import {setFormValue, formFieldBlur, validateField} from 'state/actions/formsAction'

export function mapStateToProps({forms}, {formName, name}) {
  const form = forms[formName] || {values: {}, initialValues: {}, touched: {}, validationErrors: {}}
  const value = form.values.hasOwnProperty(name) ? form.values[name] : form.initialValues[name]
  return {
    value: value,
    validationError: form.validationErrors[name] || null,
    hasError: !!(form.touched[name] && form.validationErrors[name]),
    touched: form.touched[name] || false
  }
}

export function mapDispatchToProps(dispatch, {formName, name}) {
  return {
    onChange: ({target: {value}}) => dispatch(setFormValue({form: formName, name, value})),
    onBlur: () => {
      dispatch(formFieldBlur({form: formName, name}))
    },
    validateField: (error) => dispatch(validateField({form: formName, name, error}))
  }
}

const ConnectedFormField = connect(mapStateToProps, mapDispatchToProps)(FormField)

function ConnectedFormFieldWithContext(props, context) {
  const formName = context._form && context._form.name
  return <ConnectedFormField {...props} formName={formName} />
}

ConnectedFormFieldWithContext.contextTypes = {
  _form: PropTypes.object
}

export default ConnectedFormFieldWithContext