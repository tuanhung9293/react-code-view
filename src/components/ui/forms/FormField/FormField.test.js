import React from 'react'
import {shallow, mount} from 'enzyme'
import FormField from './FormField'
import Label from '../Label'
import ErrorLabel from '../ErrorLabel'
import Input from '../Input'


describe('validation', () => {

  let onBlur, onChange, validator, validateField, registerValidator, unregisterValidator, context

  beforeEach(() => {
    onBlur = jest.fn()
    onChange = jest.fn()
    validator = jest.fn().mockImplementation(() => 'fail error')
    validateField = jest.fn()
    registerValidator = jest.fn()
    unregisterValidator = jest.fn()
    context = {
      _form: {
        registerValidator,
        unregisterValidator
      }
    }
  })

  it('registers validator to context on mount', () => {
    const wrapper = mount(<FormField name="testField" validate={validator} />, {context})
    expect(registerValidator).toBeCalledWith('testField', validator)
  })

  it('runs the validator when mounted', () => {
    jest.useFakeTimers()
    const wrapper = mount(
      <FormField
        name="testField"
        validate={validator}
        validateField={validateField}
      />, {context}
    )
    jest.runAllTimers()
    expect(validateField).toBeCalled()
    expect(validator).toBeCalled()
  })

  it('skips registering validator to context if none', () => {
    const wrapper = mount(<FormField name="testField" />, {context})
    expect(registerValidator).not.toBeCalled()
  })

  it('unregisters validator to context on unmount', () => {
    const wrapper = shallow(<FormField name="testField" validate={validator} />, {context})
    wrapper.unmount()
    expect(unregisterValidator).toBeCalledWith('testField')
  })

  it('calls onBlur on blur', () => {
    const wrapper = shallow(
      <FormField
        name="testField"
        value=""
        onBlur={onBlur}
        validate={validator}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: ''}}
    wrapper.find('Input').simulate('blur', event)
    expect(onBlur).toBeCalledWith(event)
  })

  it('calls validator on blur', () => {
    const wrapper = shallow(
      <FormField
        name="testField"
        value="test value"
        onBlur={onBlur}
        validate={validator}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('blur', event)
    expect(validator).toBeCalledWith('test value')
  })

  it('does not call validator on blur if none', () => {
    const wrapper = shallow(
      <FormField
        name="testField"
        value="test value"
        onBlur={onBlur}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('blur', event)
    expect(validator).not.toBeCalled()
  })

  it('calls validateField on blur', () => {
    const wrapper = shallow(
      <FormField
        name="testField"
        value="test value"
        onBlur={onBlur}
        validate={validator}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('blur', event)
    expect(validateField).toBeCalledWith('fail error')
  })

  it('calls validateField debounced on change', () => {
    jest.useFakeTimers()
    const wrapper = mount(
      <FormField
        name="testField"
        value="test value"
        validate={validator}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('change', event)
    jest.runAllTimers()
    expect(validateField).toBeCalledWith('fail error')
  })

  it('calls cancels debounced validate on unmount', () => {
    const wrapper = mount(
      <FormField
        name="testField"
        value="test value"
        validate={validator}
        validateField={validateField}
      />,
      {context}
    )
    const instance = wrapper.instance()
    instance.debouncedValidate = jest.fn()
    instance.debouncedValidate.cancel = jest.fn()
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.unmount()
    expect(instance.debouncedValidate.cancel).toBeCalled()
  })

  it('calls onChange handler on change if it exists', () => {
    const wrapper = shallow(
      <FormField
        name="testField"
        value="test value"
        validate={validator}
        onChange={onChange}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('change', event)
    expect(onChange).toBeCalledWith(event)
  })

  it('calls onValidationError when there is an error', () => {
    const onValidationError = jest.fn()
    const wrapper = shallow(
      <FormField
        name="testField"
        value="test value"
        onBlur={onBlur}
        onValidationError={onValidationError}
        validate={validator}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('blur', event)
    expect(onValidationError).toBeCalledWith('fail error')
  })

  it('does not call onValidationError if there is no error', () => {
    const onValidationError = jest.fn()
    const wrapper = shallow(
      <FormField
        name="testField"
        value="test value"
        onBlur={onBlur}
        onValidationError={onValidationError}
        validate={() => false}
        validateField={validateField}
      />,
      {context}
    )
    const event = {target: {name: 'testField', value: 'hej'}}
    wrapper.find('Input').simulate('blur', event)
    expect(onValidationError).toHaveBeenCalledTimes(0)
  })
})

describe('rendering', () => {
  it('contains a label', () => {
    const wrapper = shallow(<FormField name="testField" label="Test" />)
    expect(wrapper).toMatchSnapshot()
  })

  it('contains an error label when there is an error', () => {
    const wrapper = mount(<FormField name="testField" error="test error" />)
    expect(wrapper).toMatchSnapshot()
  })

  it('handles errors that are objects', () => {
    const wrapper = mount(<FormField name="testField" error={{some: 'error'}} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('handles errors that are arrays of strings', () => {
    const wrapper = mount(<FormField name="testField" error={['some error', 'some other error']} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('handles errors that are arrays of objects', () => {
    const wrapper = mount(<FormField name="testField" error={[{some: 'error'}, {some: 'other error'}]} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('contains validation error when touched', () => {
    const wrapper = shallow(<FormField name="testField" validationError="test error" touched />)
    expect(wrapper).toMatchSnapshot()
  })

  it('contains no validation error when not touched', () => {
    const wrapper = shallow(<FormField name="testField" validationError="test error" />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders an Input element with type text as default', () => {
    const wrapper = shallow(<FormField name="testField" />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a component if supplied', () => {
    const TestComponent = () => (
      <div className="Component"></div>
    )
    const wrapper = shallow(<FormField name="testField" component={TestComponent} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('passes needed props to component', () => {
    const TestComponent = () => (
      <div className="Component"></div>
    )
    const props = {
      name: 'testField',
      component: TestComponent,
      value: 'test value',
      onBlur: () => {},
      onChange: () => {}
    }
    const wrapper = shallow(<FormField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

