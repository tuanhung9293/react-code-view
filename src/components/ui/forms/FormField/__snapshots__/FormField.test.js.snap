// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering contains a label 1`] = `
<div
  className="wrapper"
>
  <div
    className="labels"
  >
    <Label>
      Test
    </Label>
    <ErrorLabel />
  </div>
  <Input
    name="testField"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
</div>
`;

exports[`rendering contains an error label when there is an error 1`] = `
<FormField
  error="test error"
  name="testField"
>
  <div
    className="wrapper"
  >
    <div
      className="labels"
    >
      <Label>
        <label
          className="base"
        />
      </Label>
      <ErrorLabel>
        <label
          className="base"
        />
      </ErrorLabel>
    </div>
    <Input
      error="test error"
      name="testField"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    >
      <input
        className="input"
        error="test error"
        name="testField"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </Input>
  </div>
</FormField>
`;

exports[`rendering contains no validation error when not touched 1`] = `
<div
  className="wrapper"
>
  <div
    className="labels"
  >
    <Label />
    <ErrorLabel />
  </div>
  <Input
    name="testField"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
</div>
`;

exports[`rendering contains validation error when touched 1`] = `
<div
  className="wrapper"
>
  <div
    className="labels"
  >
    <Label />
    <ErrorLabel>
      test error
    </ErrorLabel>
  </div>
  <Input
    name="testField"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
</div>
`;

exports[`rendering handles errors that are arrays of objects 1`] = `
<FormField
  error={
    Array [
      Object {
        "some": "error",
      },
      Object {
        "some": "other error",
      },
    ]
  }
  name="testField"
>
  <div
    className="wrapper"
  >
    <div
      className="labels"
    >
      <Label>
        <label
          className="base"
        />
      </Label>
      <ErrorLabel>
        <label
          className="base"
        />
      </ErrorLabel>
    </div>
    <Input
      error={
        Array [
          Object {
            "some": "error",
          },
          Object {
            "some": "other error",
          },
        ]
      }
      name="testField"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    >
      <input
        className="input"
        error={
          Array [
            Object {
              "some": "error",
            },
            Object {
              "some": "other error",
            },
          ]
        }
        name="testField"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </Input>
  </div>
</FormField>
`;

exports[`rendering handles errors that are arrays of strings 1`] = `
<FormField
  error={
    Array [
      "some error",
      "some other error",
    ]
  }
  name="testField"
>
  <div
    className="wrapper"
  >
    <div
      className="labels"
    >
      <Label>
        <label
          className="base"
        />
      </Label>
      <ErrorLabel>
        <label
          className="base"
        />
      </ErrorLabel>
    </div>
    <Input
      error={
        Array [
          "some error",
          "some other error",
        ]
      }
      name="testField"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    >
      <input
        className="input"
        error={
          Array [
            "some error",
            "some other error",
          ]
        }
        name="testField"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </Input>
  </div>
</FormField>
`;

exports[`rendering handles errors that are objects 1`] = `
<FormField
  error={
    Object {
      "some": "error",
    }
  }
  name="testField"
>
  <div
    className="wrapper"
  >
    <div
      className="labels"
    >
      <Label>
        <label
          className="base"
        />
      </Label>
      <ErrorLabel>
        <label
          className="base"
        />
      </ErrorLabel>
    </div>
    <Input
      error={
        Object {
          "some": "error",
        }
      }
      name="testField"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    >
      <input
        className="input"
        error={
          Object {
            "some": "error",
          }
        }
        name="testField"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </Input>
  </div>
</FormField>
`;

exports[`rendering passes needed props to component 1`] = `
<div
  className="wrapper"
>
  <div
    className="labels"
  >
    <Label />
    <ErrorLabel />
  </div>
  <TestComponent
    name="testField"
    onBlur={[Function]}
    onChange={[Function]}
    value="test value"
  />
</div>
`;

exports[`rendering renders a component if supplied 1`] = `
<div
  className="wrapper"
>
  <div
    className="labels"
  >
    <Label />
    <ErrorLabel />
  </div>
  <TestComponent
    name="testField"
    onBlur={[Function]}
    onChange={[Function]}
  />
</div>
`;

exports[`rendering renders an Input element with type text as default 1`] = `
<div
  className="wrapper"
>
  <div
    className="labels"
  >
    <Label />
    <ErrorLabel />
  </div>
  <Input
    name="testField"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
</div>
`;
